rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,554 @@
+as it currently relies on the legacy configuration to run (and is running from the legacy service). This task is currently
+planned for 7.13 or early 7.14, see [the associated GH issue](https://github.com/elastic/kibana/issues/76935)
+
+## Technical impact on our plugins codebase
+
+### Identifying things that may break
+
+- Concurrent access to the same resources
+
+Is there, for example, some part of the code that is accessing and writing files from the data folder (or anywhere else) 
+and makes the assumption that it is the sole process actually writing to that file?
+
+- Using instanceUUID as a unique Kibana process identifier
+
+Is there, for example, schedulers that are using the instanceUUID a single process id, in opposition to a single 
+Kibana instance id? Are there situations where having the same instance UUID for all the workers is going to be a problem?
+
+- Things needing to run only once per Kibana instance
+
+Is there any part of the code that needs to be executed only once in a multi-worker mode, such as initialization code, 
+or starting schedulers? 
+
+An example would be reporting's queueFactory pooling. As we want to only be running a single headless at a time per 
+Kibana instance, only one worker should have pooling enabled.
+
+### Identified required changes
+
+(Probably not exhaustive at the moment)
+
+#### Reporting
+
+We will probably want to restrict to a single headless per Kibana instance. For that, we will have to change the logic 
+in [createQueueFactory](https://github.com/elastic/kibana/blob/4584a8b570402aa07832cf3e5b520e5d2cfa7166/x-pack/plugins/reporting/server/lib/create_queue.ts#L60-L64) 
+to only have the 'main' worker be pooling for reporting tasks.
+
+#### Telemetry
+
+- Server side fetcher
+  
+The telemetry/server/fetcher.ts will attempt sending the telemetry usage multiple times once per day from each process. 
+We do store a state in the SavedObjects store of the last time the usage was sent to prevent sending multiple times 
+(although race conditions might occur).
+
+- Tasks storing telemetry data
+  
+We have tasks across several plugins storing data in savedobjects specifically for telemetry. Under clustering these 
+tasks will be registered multiple times.
+
+Note that sending the data multiple times doesn’t have any real consequences, apart from the additional number of ES requests, 
+so this should be considered non-blocking and only an improvement.
+
+- Event-based telemetry
+
+Event-based telemetry may be affected as well. Both the existing one in the Security Solutions team and the general 
+one that is in the works. More specifically, the size of the queues will be multiplied per worker, also growing in the
+amount of network bandwidth used, and potentially affecting our customers.
+
+We could address that by making sure that the queues are held only in the main worker.
+
+#### TaskManager
```

Comment:
We've started to see quite a few issues with TaskManager running in the same process as the HTTP Server. For example, there was recently a SIEM detection rule that exerted such a load on Kibana that all HTTP requests to Kibana failed.

Long-term, it'd be great to get the TaskManager in its own worker so it can't affect anything else negatively. Will the main worker only initially be responsible for task-manager and running saved-object migrations on startup? If so, it's probably fine to have the TaskManager here short-term.

Reply:
> Long-term, it'd be great to get the TaskManager in its own worker so it can't affect anything else negatively.

I'd take this one step further and say it'd also be great if a single task couldn't starve the other tasks for resources as well. I think to solve this, we'd need to use a proper pool of worker threads to execute individual tasks on.

Agreed that having task manager run on a single worker for now would get us quite a bit of value. At least we can still continue accepting requests when a CPU-intensive task is running.

Reply:
> I'd take this one step further and say it'd also be great if a single task couldn't starve the other tasks for resources as well. I think to solve this, we'd need to use a proper pool of worker threads to execute individual tasks on.

👍

Reply:
This has been clarified in the RFC. Current proposal is for TM to live in the `main` worker, which will still serve http requests, but at least TM won't be affecting other workers.

Long term, we'll look toward a strategy of having a dedicated process for background tasks to fully separate them from http requests.